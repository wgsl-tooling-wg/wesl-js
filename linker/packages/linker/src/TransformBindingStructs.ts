import { tracing } from "mini-parse";
import {
  AbstractElem,
  AttributeElem,
  BindingStructElem,
  DeclarationElem,
  FnElem,
  ModuleElem,
  StructElem,
  StructMemberElem,
  SyntheticElem,
  TypeTemplateParameter,
} from "./AbstractElems.ts";
import { declUniqueName } from "./BindIdents.ts";
import { TransformedAST } from "./Linker.ts";
import { visitAst } from "./LinkerUtil.ts";
import { findDecl } from "./LowerAndEmit.ts";
import {
  attributeToString,
  contentsToString,
  typeListToString,
  typeParamToString,
} from "./RawEmit.ts";
import { DeclIdent, RefIdent } from "./Scope.ts";
import { filterMap } from "./Util.ts";
import { textureStorage } from "./Reflection.ts";

/**
 * Transform binding structures into binding variables by mutating the AST.
 *
 * First we find all the binding structs:
 *   . find all the structs in the module by filtering the moduleElem.contents
 *     . for each struct:
 *       . mark any structs with that contain @group or @binding annotations as 'binding structs' and save them in a list
 *       . (later) create reverse links from structs to struct members
 *       . (later) visit all the binding structs and traverse to referencing structs, marking the referencing structs as binding structs too
 * Generate synethic AST nodes for binding variables
 *
 * Find all references to binding struct members
 *   . find the componound idents by traversing moduleElem.contents
 *   . filter to find the compound idents that refer to 'binding structs'
 *     . go from each ident to its declaration,
 *     . declaration to typeRef reference
 *     . typeRef to type declaration
 *     . check type declaration to see if it's a binding struct
 *     . record the intermediate declaration (e.g. a fn param b:Bindings from 'fn(b:Bindings)' )
 * rewrite references to binding struct members as synthetic elements
 *
 * Remove the binding structs from the AST
 * Remove the intermediate fn param declarations from the AST
 * Add the new binding variables to the AST
 *
 * @return the binding structs and the mutated AST
 */
export function lowerBindingStructs(ast: TransformedAST): TransformedAST {
  const { moduleElem, globalNames, notableElems } = ast;
  const bindingStructs = markBindingStructs(moduleElem); // CONSIDER should we only mark bining structs referenced from the entry point?
  markEntryTypes(moduleElem, bindingStructs);
  const newVars = bindingStructs.flatMap(s =>
    transformBindingStruct(s, globalNames),
  );
  const bindingRefs = findRefsToBindingStructs(moduleElem);

  // convert references 'b.particles' to references to the synthetic var 'particles'
  bindingRefs.forEach(({ memberRef, struct }) =>
    transformBindingReference(memberRef, struct),
  );
  // remove intermediate fn param declaration b:Bindings from 'fn(b:Bindings)'
  bindingRefs.forEach(({ intermediates }) =>
    intermediates.forEach(e => (e.contents = [])),
  );
  const contents = removeBindingStructs(moduleElem);
  moduleElem.contents = [...newVars, ...contents];
  notableElems.bindingStructs = bindingStructs;
  return { ...ast, moduleElem };
}

export function markEntryTypes(
  moduleElem: ModuleElem,
  bindingStructs: BindingStructElem[],
): void {
  const fns = moduleElem.contents.filter(e => e.kind === "fn");
  const fnFound = fnReferencesBindingStruct(fns, bindingStructs);
  if (fnFound) {
    const { fn, struct } = fnFound;
    struct.entryFn = fn;
  }
}

function fnReferencesBindingStruct(
  fns: FnElem[],
  bindingStructs: BindingStructElem[],
): { fn: FnElem; struct: BindingStructElem } | undefined {
  for (const fn of fns) {
    const { params } = fn;
    for (const p of params) {
      const ref = p.name?.typeRef?.name as RefIdent | undefined;
      const referencedElem = (ref?.refersTo as DeclIdent)
        ?.declElem as StructElem;
      const struct = bindingStructs.find(s => s === referencedElem);
      if (struct) {
        return { fn, struct };
      }
    }
  }
}

function removeBindingStructs(moduleElem: ModuleElem): AbstractElem[] {
  return moduleElem.contents.filter(
    elem => elem.kind !== "struct" || !elem.bindingStruct,
  );
}

/** mutate the AST, marking StructElems as bindingStructs
 *  (if they contain ptrs with @group @binding annotations)
 * @return the binding structs
 */
export function markBindingStructs(
  moduleElem: ModuleElem,
): BindingStructElem[] {
  const structs = moduleElem.contents.filter(elem => elem.kind === "struct");
  const bindingStructs = structs.filter(containsBinding);
  bindingStructs.forEach(struct => (struct.bindingStruct = true));
  // LATER also mark structs that reference a binding struct..
  return bindingStructs as BindingStructElem[];
}

/** @return true if this struct contains a member with marked with @binding or @group */
function containsBinding(struct: StructElem): boolean {
  return struct.members.some(({ attributes }) => bindingAttribute(attributes));
}

function bindingAttribute(attributes?: AttributeElem[]): boolean {
  if (!attributes) return false;
  return attributes.some(({ name }) => name === "binding" || name === "group");
}

/** convert each member of the binding struct into a synthetic global variable */
export function transformBindingStruct(
  s: StructElem,
  globalNames: Set<string>,
): SyntheticElem[] {
  return s.members.map(member => {
    const { typeRef, name: memberName } = member;
    const { name: typeName } = typeRef!; // members should always have a typeRef.. TODO fix typing to show this
    const typeParameters = typeRef?.templateParams;

    const varName = declUniqueName(memberName.name, globalNames);
    member.mangledVarName = varName; // save new name so we can rewrite references to this member later
    globalNames.add(varName);

    const attributes =
      member.attributes?.map(attributeToString).join(" ") ?? "";
    const varTypes =
      lowerPtrMember(member, typeName, typeParameters, varName) ??
      lowerStdTypeMember(typeName, typeParameters) ??
      lowerStorageTextureMember(typeName, typeParameters);
    if (!varTypes) {
      console.log("unhandled case transforming member", typeName);
      return syntheticVar(attributes, varName, "", "??");
    }

    const { storage: storageType, varType } = varTypes;
    return syntheticVar(attributes, varName, storageType, varType);
  });
}

interface LoweredVarTypes {
  storage: string;
  varType: string;
}

function lowerPtrMember(
  member: StructMemberElem,
  typeName: string | RefIdent,
  typeParameters: TypeTemplateParameter[] | undefined,
  varName: string,
): LoweredVarTypes | undefined {
  if (typeName === "ptr") {
    const origParams = typeParameters ?? [];
    const newParams = [origParams[0]];
    if (origParams[2]) newParams.push(origParams[2]);
    const storage = typeListToString(newParams);

    const varType = typeParamToString(origParams?.[1]);
    return { storage, varType };
  }
}

function lowerStdTypeMember(
  typeName: string | RefIdent,
  typeParameters: TypeTemplateParameter[] | undefined,
): LoweredVarTypes | undefined {
  if (typeof typeName !== "string") {
    const varBaseType = typeName.std ? typeName.originalName : "??";
    const params = typeParameters ? typeListToString(typeParameters) : "";
    const varType = varBaseType + params;

    return { varType, storage: "" };
  }
}

function lowerStorageTextureMember(
  typeName: string | RefIdent,
  typeParameters: TypeTemplateParameter[] | undefined,
): LoweredVarTypes | undefined {
  if (typeof typeName === "string" && textureStorage.test(typeName)) {
    const params = typeParameters ? typeListToString(typeParameters) : "";
    const varType = typeName + params;
    return { varType, storage: "" };
  }
}

function syntheticVar(
  attributes: string,
  varName: string,
  storageTemplate: string,
  varType: string,
): SyntheticElem {
  const varText = `var${storageTemplate} ${attributes} ${varName} : ${varType};\n`;

  const elem: SyntheticElem = {
    kind: "synthetic",
    text: varText,
  };
  return elem;
}

interface MemberRefToStruct extends StructTrace {
  memberRef: SimpleMemberRef; // e.g. the memberRef 'b.particles'
}

interface StructTrace {
  struct: StructElem; // e.g. the struct Bindings
  intermediates: DeclarationElem[]; // e.g. the fn param b:Bindings from 'fn(b:Bindings)'
}

/** find all simple member references in the module that refer to binding structs */
export function findRefsToBindingStructs(
  moduleElem: ModuleElem,
): MemberRefToStruct[] {
  const members: SimpleMemberRef[] = [];
  visitAst(moduleElem, elem => {
    if (elem.kind === "memberRef") members.push(elem);
  });
  return filterMap(members, refersToBindingStruct);
}

/** @return true if this memberRef refers to a binding struct */
function refersToBindingStruct(
  memberRef: SimpleMemberRef,
): MemberRefToStruct | undefined {
  const found = traceToStruct(memberRef.name.ident);

  if (found && found.struct.bindingStruct) {
    return { memberRef, ...found };
  }
}

/** If this identifier ultimately refers to a struct type, return the struct declaration */
function traceToStruct(ident: RefIdent): StructTrace | undefined {
  const decl = findDecl(ident);
  const declElem = decl.declElem;
  // for now only handle the case where the reference points at a fn parameter
  if (declElem.kind === "param") {
    const name = declElem.name.typeRef!.name;
    if (typeof name !== "string") {
      if (name.std) {
        return undefined;
      }

      const paramDecl = findDecl(name);
      const structElem = paramDecl.declElem;
      if (structElem.kind === "struct") {
        return { struct: structElem, intermediates: [declElem] };
      }
      return undefined;
    }
  } else {
    // LATER presumably handle other cases? Should this be more general, e.g. traceToType()?
    // elemLog(
    //   ident.refIdentElem!,
    //   `unhandled case in traceToStruct: decl ${declElem.kind} not yet implemented`,
    // );
  }
}

/** Mutate the member reference elem to instead contain synthetic elem text.
 * The new text is the mangled var name of the struct member that the memberRef refers to. */
export function transformBindingReference(
  memberRef: SimpleMemberRef,
  struct: StructElem,
): SyntheticElem {
  const refName = memberRef.member.name;
  const structMember = struct.members.find(m => m.name.name === refName)!;
  if (!structMember || !structMember.mangledVarName) {
    if (tracing) console.log(`missing mangledVarName for ${refName}`);
    return { kind: "synthetic", text: refName };
  }
  const { extraComponents } = memberRef;
  const extraText = extraComponents ? contentsToString(extraComponents) : "";

  const text = structMember.mangledVarName + extraText;
  const synthElem: SyntheticElem = { kind: "synthetic", text };
  memberRef.contents = [synthElem];
  return synthElem;
}
