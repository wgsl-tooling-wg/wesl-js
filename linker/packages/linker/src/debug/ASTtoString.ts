import {
  AbstractElem,
  ExpressionElem,
  ModuleElem,
  TypeRefElem,
  TypeTemplateParameter,
} from "../AbstractElems.ts";
import { importToString } from "./ImportToString.ts";
import { LineWrapper } from "./LineWrapper.ts";

const maxLineLength = 150;

export function astToString(elem: AbstractElem, indent = 0): string {
  const { kind, contents } = elem as ModuleElem;
  const str = new LineWrapper(indent, maxLineLength);
  str.add(kind);
  addElemFields(elem, str);
  let childStrings: string[] = [];
  if (contents) {
    childStrings = contents.map(e => astToString(e, indent + 2));
  }
  if (childStrings.length) {
    str.nl();
    str.addBlock(childStrings.join("\n"), false);
  }

  return str.result;
}

// TODO DRY with above
export function elemToString(elem: AbstractElem): string {
  const { kind } = elem as ModuleElem;
  const str = new LineWrapper(0, maxLineLength);
  str.add(kind);
  addElemFields(elem, str);
  let childStrings: string[] = [];
  if (childStrings.length) {
    str.nl();
    str.addBlock(childStrings.join("\n"), false);
  }
  return str.result;
}

function addElemFields(elem: AbstractElem, str: LineWrapper): void {
  addTextFields(elem, str) ||
    addVarishFields(elem, str) ||
    addStructFields(elem, str) ||
    addStructMemberFields(elem, str) ||
    addNameFields(elem, str) ||
    addMemberRef(elem, str) ||
    addFnFields(elem, str) ||
    addAliasFields(elem, str) ||
    addAttributeFields(elem, str) ||
    addExpressionFields(elem, str) ||
    addTypeRefFields(elem, str) ||
    addSynthetic(elem, str) ||
    addImport(elem, str) ||
    addRefIdent(elem, str) ||
    addTypedDeclIdent(elem, str) ||
    addDeclIdent(elem, str);
}

function addVarishFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  const { kind } = elem;
  if (
    kind === "var" ||
    kind === "let" ||
    kind === "gvar" ||
    kind === "const" ||
    kind === "override"
  ) {
    addTypedDeclIdent(elem.name, str);
    return true;
  }
}

function addTextFields(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "text") {
    const { srcModule, start, end } = elem;
    str.add(` '${srcModule.src.slice(start, end)}'`);
    return true;
  }
}

function addRefIdent(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "ref") {
    str.add(" " + elem.ident.originalName);
    return true;
  }
}

function addMemberRef(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "memberRef") {
    str.add(` ${elem.name.ident.originalName}.${elem.member.name}`);
    return true;
  }
}

function addDeclIdent(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "decl") {
    const { ident } = elem;
    str.add(" %" + ident.originalName);
    return true;
  }
}

function addTypedDeclIdent(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "typeDecl") {
    const { decl, typeRef } = elem;
    str.add(" %" + decl.ident.originalName);
    if (typeRef) {
      str.add(" : " + typeRefElemToString(typeRef));
    }
    return true;
  }
}

function addAliasFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "alias") {
    const { name, typeRef } = elem;
    const prefix = name.ident.kind === "decl" ? "%" : "";
    str.add(" " + prefix + name.ident.originalName);
    str.add("=" + typeRefElemToString(typeRef));
    return true;
  }
}

function addStructFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "struct") {
    str.add(" " + elem.name.ident.originalName);
    return true;
  }
}

function addStructMemberFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "member") {
    const { name, typeRef, attributes } = elem;
    if (attributes) {
      str.add(" " + attributes.map(a => "@" + (a && a.name)).join(" "));
    }
    str.add(" " + name.name);
    str.add(": " + typeRefElemToString(typeRef));
    return true;
  }
}

function addImport(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "import") {
    str.add(" " + importToString(elem.imports));
    return true;
  }
}

function addSynthetic(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "synthetic") {
    str.add(` '${elem.text}'`);
    return true;
  }
}

function addAttributeFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "attribute") {
    const { name, params } = elem;
    str.add(" @" + name);
    if (params) {
      str.add("(");
      str.add(params.map(expressionToString).join(", "));
      str.add(")");
    }
    return true;
  }
}

function addNameFields(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "name") {
    str.add(" " + elem.name);
    return true;
  }
}

function addFnFields(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "fn") {
    const { name, params, returnType, fnAttributes } = elem;

    str.add(" " + name.ident.originalName);

    str.add("(");
    const paramStrs = params
      .map(
        (
          p, // TODO DRY
        ) => {
          const { name } = p;
          const { originalName } = name.decl.ident;
          const typeRef = typeRefElemToString(name.typeRef!);
          return originalName + ": " + typeRef;
        },
      )
      .join(", ");
    str.add(paramStrs);
    str.add(")");

    fnAttributes?.forEach(a => {
      str.add(" @" + a?.name);
      if (a?.params) {
        str.add("(");
        str.add(a.params.map(expressionToString).join(", "));
        str.add(")");
      }
    });

    if (returnType) {
      str.add(" -> " + typeRefElemToString(returnType));
    }

    return true;
  }
}

function addTypeRefFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "type") {
    const { name } = elem;
    const nameStr = typeof name === "string" ? name : name.originalName;
    str.add(" " + nameStr);

    if (elem.templateParams !== undefined) {
      const paramStrs = elem.templateParams
        .map(templateParamToString)
        .join(", ");
      str.add("<" + paramStrs + ">");
    }
    return true;
  }
}
function addExpressionFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "expression") {
    const contents = elem.contents
      .map(e => {
        if (e.kind === "text") {
          return "'" + e.srcModule.src.slice(e.start, e.end) + "'";
        } else {
          return elemToString(e);
        }
      })
      .join(" ");
    str.add(" " + contents);
    return true;
  }
}

function expressionToString(elem: ExpressionElem): string {
  const contents = elem.contents
    .map(e => {
      if (e.kind === "text") {
        return "'" + e.srcModule.src.slice(e.start, e.end) + "'";
      } else {
        return elemToString(e);
      }
    })
    .join(" ");
  return contents;
}

function templateParamToString(p: TypeTemplateParameter): string {
  if (typeof p === "string") {
    return p;
  } else if (p.kind === "type") {
    return typeRefElemToString(p);
  } else if (p.kind === "expression") {
    return expressionToString(p);
  } else {
    console.log("unknown template parameter type", p);
    return "??";
  }
}

function typeRefElemToString(elem: TypeRefElem): string {
  if (!elem) return "?type?";
  const { name } = elem;
  const nameStr = typeof name === "string" ? name : name.originalName;

  let params = "";
  if (elem.templateParams !== undefined) {
    const paramStrs = elem.templateParams.map(templateParamToString).join(", ");
    params = "<" + paramStrs + ">";
  }
  return nameStr + params;
}
