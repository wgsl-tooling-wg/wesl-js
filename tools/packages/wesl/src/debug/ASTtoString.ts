import {
  assertThat,
  assertUnreachable,
} from "../../../mini-parse/src/Assertions.ts";
import {
  AbstractElem,
  AttributeElem,
  ContainerElem,
  ExpressionElem,
  FnElem,
  StuffElem,
  TranslateTimeExpressionElem,
  TypedDeclElem,
  TypeRefElem,
  TypeTemplateParameter,
  UnknownExpressionElem,
} from "../AbstractElems.ts";
import { importToString } from "./ImportToString.ts";
import { LineWrapper } from "./LineWrapper.ts";

const maxLineLength = 150;

export function astToString(elem: AbstractElem, indent = 0): string {
  const { kind } = elem;
  const str = new LineWrapper(indent, maxLineLength);
  str.add(kind);
  addElemFields(elem, str);
  let childStrings: string[] = [];
  if ("contents" in elem) {
    childStrings = elem.contents.map(e => astToString(e, indent + 2));
  }
  if (childStrings.length) {
    str.nl();
    str.addBlock(childStrings.join("\n"), false);
  }

  return str.result;
}

function addElemFields(elem: AbstractElem, str: LineWrapper): void {
  const { kind } = elem;
  if (kind === "text") {
    const { srcModule, start, end } = elem;
    str.add(` '${srcModule.src.slice(start, end)}'`);
  } else if (
    kind === "var" ||
    kind === "let" ||
    kind === "gvar" ||
    kind === "const" ||
    kind === "override"
  ) {
    addTypedDeclIdent(elem.name, str);
  } else if (kind === "struct") {
    str.add(" " + elem.name.ident.originalName);
  } else if (kind === "member") {
    const { name, typeRef, attributes } = elem;
    if (attributes) {
      str.add(" " + attributes.map(a => "@" + a.name).join(" "));
    }
    str.add(" " + name.name);
    str.add(": " + typeRefElemToString(typeRef));
  } else if (kind === "name") {
    str.add(" " + elem.name);
  } else if (kind === "memberRef") {
    const { extraComponents } = elem;
    const extraText =
      extraComponents ? debugContentsToString(extraComponents) : "";
    str.add(` ${elem.name.ident.originalName}.${elem.member.name}${extraText}`);
  } else if (kind === "fn") {
    addFnFields(elem, str);
  } else if (kind === "alias") {
    const { name, typeRef } = elem;
    const prefix = name.ident.kind === "decl" ? "%" : "";
    str.add(" " + prefix + name.ident.originalName);
    str.add("=" + typeRefElemToString(typeRef));
  } else if (kind === "attribute") {
    addAttribute(elem, str);
  } else if (kind === "expression") {
    const contents = elem.contents
      .map(e => {
        if (e.kind === "text") {
          return "'" + e.srcModule.src.slice(e.start, e.end) + "'";
        } else {
          return astToString(e);
        }
      })
      .join(" ");
    str.add(" " + contents);
  } else if (kind === "type") {
    const { name } = elem;
    const nameStr = typeof name === "string" ? name : name.originalName;
    str.add(" " + nameStr);

    if (elem.templateParams !== undefined) {
      const paramStrs = elem.templateParams
        .map(templateParamToString)
        .join(", ");
      str.add("<" + paramStrs + ">");
    }
  } else if (kind === "synthetic") {
    str.add(` '${elem.text}'`);
  } else if (kind === "import") {
    str.add(" " + importToString(elem.imports));
  } else if (kind === "ref") {
    str.add(" " + elem.ident.originalName);
  } else if (kind === "typeDecl") {
    addTypedDeclIdent(elem, str);
  } else if (kind === "decl") {
    const { ident } = elem;
    str.add(" %" + ident.originalName);
  } else if (kind === "assert") {
    // Nothing to do for now
  } else if (kind === "module") {
    // Ignore this kind of elem
  } else if (kind === "param") {
    // TODO: This branch shouldn't exist
  } else if (kind === "stuff") {
    // Ignore
  } else if (kind === "translate-time-expression") {
    addTranslateTimeExpression(elem, str);
  } else if (kind === "directive") {
    str.add(` ${elem.directive}${elem.arguments.join(",")}`);
  } else {
    assertUnreachable(kind);
  }
}

function addAttribute(elem: AttributeElem, str: LineWrapper) {
  const { name, params } = elem;
  str.add(" @" + name);
  if (name === "if") {
    assertThat(params !== undefined);
    assertThat(params.every(p => p.kind === "translate-time-expression"));
    str.add("(");
    str.add(params.map(v => expressionToString(v.expression)).join(", "));
    str.add(")");
  } else {
    if (params) {
      assertThat(params.every(p => p.kind === "expression"));
      str.add("(");
      str.add(params.map(unknownExpressionToString).join(", "));
      str.add(")");
    }
  }
}

function addTypedDeclIdent(elem: TypedDeclElem, str: LineWrapper) {
  const { decl, typeRef } = elem;
  str.add(" %" + decl.ident.originalName);
  if (typeRef) {
    str.add(" : " + typeRefElemToString(typeRef));
  }
}

function addFnFields(elem: FnElem, str: LineWrapper) {
  const { name, params, returnType, fnAttributes } = elem;

  str.add(" " + name.ident.originalName);

  str.add("(");
  const paramStrs = params
    .map(
      (
        p, // TODO DRY
      ) => {
        const { name } = p;
        const { originalName } = name.decl.ident;
        const typeRef = typeRefElemToString(name.typeRef!);
        return originalName + ": " + typeRef;
      },
    )
    .join(", ");
  str.add(paramStrs);
  str.add(")");

  fnAttributes?.forEach(a => addAttribute(a, str));

  if (returnType) {
    str.add(" -> " + typeRefElemToString(returnType));
  }
}

function addTranslateTimeExpression(
  elem: TranslateTimeExpressionElem,
  str: LineWrapper,
) {
  str.add(expressionToString(elem.expression));
}

function expressionToString(elem: ExpressionElem): string {
  const { kind } = elem;
  if (kind === "binary-expression") {
    return `${expressionToString(elem.left)} ${elem.operator.value} ${expressionToString(elem.right)}`;
  } else if (kind === "unary-expression") {
    return `${elem.operator.value}${expressionToString(elem.expression)}`;
  } else if (kind === "ref") {
    return elem.ident.originalName;
  } else if (kind === "literal") {
    return elem.value;
  } else if (kind === "translate-time-feature") {
    return elem.name;
  } else if (kind === "parenthesized-expression") {
    return `(${expressionToString(elem.expression)})`;
  } else if (kind === "component-expression") {
    return `${expressionToString(elem.base)}[${elem.access}]`;
  } else if (kind === "component-member-expression") {
    return `${expressionToString(elem.base)}.${elem.access}`;
  } else if (kind === "call-expression") {
    return `${elem.function.ident.originalName}(${elem.arguments.map(expressionToString).join(", ")})`;
  } else {
    assertUnreachable(kind);
  }
}

function unknownExpressionToString(elem: UnknownExpressionElem): string {
  // TODO: Temp hack while I clean up the expression parsing
  if ("contents" in elem) {
    // @ts-ignore
    const contents = elem.contents
      // @ts-ignore
      .map(e => {
        if (e.kind === "text") {
          return "'" + e.srcModule.src.slice(e.start, e.end) + "'";
        } else {
          return astToString(e);
        }
      })
      .join(" ");
    return contents;
  }
  return astToString(elem);
}

function templateParamToString(p: TypeTemplateParameter): string {
  if (typeof p === "string") {
    return p;
  } else if (p.kind === "type") {
    return typeRefElemToString(p);
  } else if (p.kind === "expression") {
    return unknownExpressionToString(p);
  } else {
    console.log("unknown template parameter type", p);
    return "??";
  }
}

function typeRefElemToString(elem: TypeRefElem): string {
  if (!elem) return "?type?";
  const { name } = elem;
  const nameStr = typeof name === "string" ? name : name.originalName;

  let params = "";
  if (elem.templateParams !== undefined) {
    const paramStrs = elem.templateParams.map(templateParamToString).join(", ");
    params = "<" + paramStrs + ">";
  }
  return nameStr + params;
}

export function debugContentsToString(elem: StuffElem): string {
  const parts = elem.contents.map(c => {
    const { kind } = c;
    if (kind === "text") {
      return c.srcModule.src.slice(c.start, c.end);
    } else if (kind === "ref") {
      return c.ident.originalName; // not using the mapped to decl name, so this can be used for debug..
    } else {
      return `?${c.kind}?`;
    }
  });
  return parts.join(" ");
}
