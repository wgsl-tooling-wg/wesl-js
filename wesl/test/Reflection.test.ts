import { expect } from "@std/expect";
import type { BindingStructElem, StructElem } from "../AbstractElems.ts";
import { astToString } from "../debug/ASTtoString.ts";
import {
  bindingGroupLayoutTs,
  reportBindingStructsPlugin,
} from "../Reflection.ts";
import { bindingStructsPlugin } from "../TransformBindingStructs.ts";
import { linkTestOpts } from "./TestUtil.ts";
import { assertSnapshot } from "@std/testing/snapshot";

Deno.test("extract binding struct", async (t) => {
  const src = `
    struct Bindings {
      @group(0) @binding(0) particles: ptr<storage, array<f32>, read_write>, 
    }

    @compute fn main(b: Bindings) {
      let x = b.particles;
    }
  `;
  let found: StructElem[] | undefined;
  const config = {
    plugins: [
      bindingStructsPlugin(),
      reportBindingStructsPlugin((report) => (found = report)),
    ],
  };
  linkTestOpts({ config }, src);

  // verify struct found
  expect(found).toBeDefined();
  const s = found![0];
  expect(s).toBeDefined();
  await assertSnapshot(t, astToString(s));
  expect(s.bindingStruct).toBeTruthy();
  expect((s as BindingStructElem).entryFn).toBeDefined();

  // verify struct members
  const members = s.members.filter((e) => e.kind === "member");
  const membersAst = members.map((e) => astToString(e)).join("\n");
  await assertSnapshot(t, membersAst);
});

Deno.test("binding struct to ts", async (t) => {
  const src = `
    struct Uniforms {
      foo: u32
    }
    struct MyBindings {
      @group(0) @binding(0) particles: ptr<storage, array<f32>, read_write>, 
      @group(0) @binding(1) uniforms: ptr<uniform, Uniforms>, 
      @group(0) @binding(2) tex: texture_2d<f32>,
      @group(0) @binding(3) samp: sampler,
      @group(0) @binding(4) stTex: texture_storage_2d<rgba8unorm, read>,
    }
    @compute fn main(b: MyBindings) {
      let x = b.particles;
    }
  `;
  let found: StructElem[] | undefined;
  const config = {
    plugins: [
      bindingStructsPlugin(),
      reportBindingStructsPlugin((report) => (found = report)),
    ],
  };
  linkTestOpts({ config }, src);
  const ts = bindingGroupLayoutTs(found![0] as BindingStructElem);
  await assertSnapshot(t, ts);
});
