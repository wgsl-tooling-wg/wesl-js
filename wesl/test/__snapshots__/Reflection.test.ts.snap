export const snapshot = {};

snapshot[`extract binding struct 1`] = `
"struct Bindings
  text 'struct '
  decl %Bindings
  text ' {
      '
  member @group @binding particles: ptr<storage, array<f32>, read_write>
    attribute @group('0')
      expression '0'
        text '0'
    text ' '
    attribute @binding('0')
      expression '0'
        text '0'
    text ' '
    name particles
    text ': '
    type ptr<storage, array<f32>, read_write>
      ref ptr
      text '<'
      type storage
        ref storage
      text ', '
      type array<f32>
        ref array
        text '<'
        type f32
          ref f32
        text '>'
      text ', '
      type read_write
        ref read_write
      text '>'
  text ', 
    }'"
`;

snapshot[`extract binding struct 2`] = `
"member @group @binding particles: ptr<storage, array<f32>, read_write>
  attribute @group('0')
    expression '0'
      text '0'
  text ' '
  attribute @binding('0')
    expression '0'
      text '0'
  text ' '
  name particles
  text ': '
  type ptr<storage, array<f32>, read_write>
    ref ptr
    text '<'
    type storage
      ref storage
    text ', '
    type array<f32>
      ref array
      text '<'
      type f32
        ref f32
      text '>'
    text ', '
    type read_write
      ref read_write
    text '>'"
`;

snapshot[`binding struct to ts 1`] = `
'
const myBindingsEntries = [ 
      {
        binding: 0,
        visibility: GPUShaderStage.COMPUTE,
        buffer: { type: "storage" }
      },
      {
        binding: 1,
        visibility: GPUShaderStage.COMPUTE,
        buffer: { type: "uniform" }
      },
      {
        binding: 2,
        visibility: GPUShaderStage.COMPUTE,
        texture: { sampleType: "float" }
      },
      {
        binding: 3,
        visibility: GPUShaderStage.COMPUTE,
        sampler: { type: "filtering" }
      },
      {
        binding: 4,
        visibility: GPUShaderStage.COMPUTE,
        storageTexture: { format: "rgba8unorm", sampleType: "float", access: "read-only" }
      } ];
function myBindingsLayout(device: GPUDevice): GPUBindGroupLayout {
  return device.createBindGroupLayout({
    entries: myBindingsEntries 
  });
}

export const layoutFunctions = { myBindingsLayout };
export const layouts = { myBindingsEntries };
  '
`;
